packages/custom-objects-exporter/src/main.js

```diff
@@ -0,0 +1,117 @@
+/* @flow */
+import { createClient } from '@commercetools/sdk-client'
+import { createRequestBuilder } from '@commercetools/api-request-builder'
+import { createHttpMiddleware } from '@commercetools/sdk-middleware-http'
+import {
+  createAuthMiddlewareForClientCredentialsFlow,
+  createAuthMiddlewareWithExistingToken,
+} from '@commercetools/sdk-middleware-auth'
+import { createUserAgentMiddleware } from '@commercetools/sdk-middleware-user-agent'
+import JSONStream from 'JSONStream'
+import type {
+  ApiConfigOptions,
+  ExporterOptions,
+  LoggerOptions,
+} from '../../../types/customObjects'
+import type { Client, ClientRequest } from '../../../types/sdk'
+import pkg from '../package.json'
+
+export default class CustomObjectsExporter {
+  // Set type annotations
+  apiConfig: ApiConfigOptions
+  client: Client
+  logger: LoggerOptions
+  predicate: ?string
+
+  constructor(options: ExporterOptions, logger: LoggerOptions) {
+    if (!options.apiConfig)
+      throw new Error('The constructor must be passed an `apiConfig` object')
+    this.apiConfig = options.apiConfig
+    this.client = createClient({
+      middlewares: [
+        createAuthMiddlewareWithExistingToken(
+          options.accessToken ? `Bearer ${options.accessToken}` : ''
+        ),
+        createAuthMiddlewareForClientCredentialsFlow(this.apiConfig),
+        createUserAgentMiddleware({
+          libraryName: pkg.name,
+          libraryVersion: pkg.version,
+        }),
+        createHttpMiddleware({ host: this.apiConfig.apiUrl }),
+      ],
+    })
+
+    this.predicate = options.predicate
+
+    this.logger = {
+      error: () => {},
+      warn: () => {},
+      info: () => {},
+      verbose: () => {},
+      ...logger,
+    }
+  }
+
+  run(outputStream: stream$Writable) {
+    this.logger.info('Starting Export')
+    const jsonStream = JSONStream.stringify()
+    jsonStream.pipe(outputStream)
+    CustomObjectsExporter.handleOutput(outputStream, jsonStream, this)
+  }
+
+  static handleOutput(
+    outputStream: stream$Writable,
+    pipeStream: stream$Writable,
+    instance: Object
+  ) {
+    CustomObjectsExporter.fetchObjects(pipeStream, instance)
+      .then(() => {
+        instance.logger.info('Export operation completed successfully')
+        if (outputStream !== process.stdout) pipeStream.end()
+      })
+      .catch((e: Error) => {
+        outputStream.emit('error', e)
+      })
+  }
+
+  static fetchObjects(output: stream$Writable, instance: Object): Promise<any> {
```

Comment:
What's `instance` here? I got a bit confused by it being the `instance` of this class or something?